UI: Recent allocs + job allocs view

DingoEatingFuzz opened the pull request:
In an effort to make it faster to get to allocations, this brings two new features to the job detail page.

1. Recent allocations on the overview tab
2. An allocations tab that lists all allocations across all task groups

This brings the UI job detail view into better alignment with the CLI.

![image](https://user-images.githubusercontent.com/174740/43150817-b86e8e20-8f1f-11e8-91b6-888aa8f99734.png)

![image](https://user-images.githubusercontent.com/174740/43150884-e1dd9f94-8f1f-11e8-93b9-2c14d4c5c77e.png)

johncowen commented on ui/app/templates/clients/client.hbs:
```diff
@@ -101,7 +101,7 @@
               {{#t.sort-by prop="shortId"}}ID{{/t.sort-by}}
               {{#t.sort-by prop="modifyIndex" title="Modify Index"}}Modified{{/t.sort-by}}
-              {{#t.sort-by prop="name"}}Name{{/t.sort-by}}
+              {{#t.sort-by prop="createIndex" title="Create Index"}}Created{{/t.sort-by}}
```
Not a lot of background around Nomad specifically, are names not useful for folks?

DingoEatingFuzz replied:
Not really. Names for allocations are a function of task group name and count. They end up looking like `task-group-name.[3]`.

johncowen commented on ui/app/components/job-page/parts/recent-allocations.js:
```diff
@@ -0,0 +1,24 @@
+          .slice(0, 5)
+      ),
+    });
+  }),
```
Just about of interest, what's the reason for the PromiseArray? I'm guessing the allocations might not be loaded yet when you ask to sort them or something? If you have a sec would be good to get an understanding, no prob if not.

I had a quick look at some other sorting you are doing and none of them use this, are allocations different or is this something you are going to be moving to across the board?

johncowen replied:
Actually just noticed also, looks like job is an attribute on the component `(job="job")` and then you are reaching up into the job to get the allocations. Just an offshoot, would an alternative be `allocations={{job.allocations}}` so you are sending the data down? Not sure if that is workable for you or not

DingoEatingFuzz replied:
If it isn't clear, the `PromiseArray` object makes it so the returned value always has an array-like interface before and after the promise resolves.

This isn't important elsewhere because `RecordArray` (either from `store.findAll` or `model.get('asyncHasMany')` are used in templates.

Here, additional transformations need to happen (sortBy, reverse, slice). 

If the promise isn't resolved yet, this breaks because it's not an array yet. Simply doing `this.get('job.allocations').then(allocations => allocations.transformations)` also doesn't work because the computed property now returns a promise. The template can't do anything with the promise and the computed property doesn't update when the promise resolves.

By using a PromiseArray, I can do the promise control flow while also ensuring an Array interface as well as a template re-render when the promise resolves.

DingoEatingFuzz replied:
I could use `allocations=job.allocations`, and that may be better. Specifically because of the principle of least authority. I chose to use `job=job` to make all the various `job-page/parts` components feel similar. 

I don't think this violates data-down actions-up because it isn't reaching up into a higher viewer (e.g., `allocations: alias('parentView.job.allocations')`). Instead, a larger than necessary piece of data is being provided to the component and the component is reaching _down_ in that data. 

I'm willing to talk about this more if you wish. I know it's pretty nuanced.

johncowen replied:
Hey, thanks for the further detail.

> I'm willing to talk about this more if you wish. I know it's pretty nuanced.

It's fine I'm not fussed, just thought I'd highlight a alternative approach if it helps. If you decide to go that route at some point and you want to chat then quite happy to, no prob if not.

Back to the PR - Ping me when you tweak the search thing so you can undo your search when there are no results, and I can give it another once over.

Thanks,

johncowen commented:
Massive aside here, but I noticed this:

![screen shot 2018-07-27 at 11 58 06](https://user-images.githubusercontent.com/554604/43317471-d5c34bec-9194-11e8-8acc-87a87646eebe.png)

..and I'd not noticed them before when running your tests. I tried on master also and they are there aswell so maybe I've just not noticed them. assuming you get these also? One of them is related to allocations somehow, don't know if they effect anything or not but thought I'd mention in case. 

I also get a tonne of deprecation notices from plugins/addons when running the Consul tests, so no biggie if they aren't affecting anything.

DingoEatingFuzz commented:
I'm aware of those warnings in tests. They are all benign, but I should eventually clean them up.

Always appreciate your thoroughness in reviews!

DingoEatingFuzz commented:
@johncowen Good catch on the search box disappearing when the empty message is shown. I had all the code in place to keep it around, but then used the wrong variable in the `#if` 🙃 

I fixed that and added a test for it. Everything should be in good shape now.
